File: src/lib/telemetry.ts
import type { Result } from '@atomicloud/diene.result';
import { type PortError, portError } from './error.js';
import { accepted, type Checked, rejected, resultFromChecked } from './validation.js';

type TelemetryPort = 'logging' | 'metrics';
type TelemetryAttributeValue = boolean | number | string;
type TelemetryAttributes = Readonly<Record<string, TelemetryAttributeValue>>;

function checkTelemetryAttributes<P extends TelemetryPort>(
  attributes: TelemetryAttributes | undefined,
  port: P,
  operation: string,
): Checked<TelemetryAttributes | undefined, PortError<P>> {
  if (attributes === undefined) {
    return accepted(undefined);
  }
  if (
    typeof attributes !== 'object' ||
    attributes === null ||
    Array.isArray(attributes) ||
    ![Object.prototype, null].includes(Object.getPrototypeOf(attributes))
  ) {
    return rejected(portError(port, 'invalid-input', operation, 'Telemetry attributes must be a primitive record'));
  }
  const entries = Object.entries(attributes);
  if (entries.some(([key]) => key.trim() === '' || key.includes('\0'))) {
    return rejected(
      portError(port, 'invalid-input', operation, 'Telemetry attribute names must be non-blank and NUL-free'),
    );
  }
  if (
    entries.some(([, value]) => {
      const valueType = typeof value;
      return (
        (valueType !== 'string' && valueType !== 'number' && valueType !== 'boolean') ||
        (valueType === 'number' && !Number.isFinite(value))
      );
    })
  ) {
    return rejected(
      portError(port, 'invalid-input', operation, 'Telemetry attribute values must be finite primitives'),
    );
  }
  return accepted(
    Object.freeze(Object.fromEntries(entries.sort(([left], [right]) => (left === right ? 0 : left < right ? -1 : 1)))),
  );
}

function validateTelemetryAttributes<P extends TelemetryPort>(
  attributes: TelemetryAttributes | undefined,
  port: P,
  operation: string,
): Result<TelemetryAttributes | undefined, PortError<P>> {
  return resultFromChecked(checkTelemetryAttributes(attributes, port, operation));
}

export type { TelemetryAttributes, TelemetryAttributeValue };
export { checkTelemetryAttributes, validateTelemetryAttributes };

File: packages/diene_interfaces/lib/src/telemetry.dart
/// Shared attribute rules for the logging and metrics emit seams.
///
/// Both telemetry seams carry the same attribute bag, so the validation lives
/// here once instead of twice. Dart/Flutter telemetry rides Faro at runtime —
/// this library ships no exporter and no OTel implementer (RB-19).
library;

import 'dart:collection';

import 'package:diene_result/diene_result.dart';

import 'port_problem.dart';

/// A validated telemetry attribute bag.
///
/// Values are restricted to finite primitives so any downstream transport
/// (Faro, a log formatter, a metrics registry) can render them without
/// negotiating a schema.
typedef TelemetryAttributes = Map<String, Object?>;

/// The empty, unmodifiable attribute bag.
final TelemetryAttributes emptyTelemetryAttributes =
    UnmodifiableMapView<String, Object?>(const <String, Object?>{});

/// Validates and canonicalises a telemetry attribute bag.
///
/// Accepts a `null` bag as the empty bag. Rejects blank or NUL-bearing keys and
/// any value that is not a finite `num`, `bool`, or `String`. On success the
/// returned map is key-sorted and unmodifiable, so two equal bags always render
/// identically.
Result<TelemetryAttributes> checkTelemetryAttributes(
  TelemetryAttributes? attributes, {
  required PortName port,
  required String operation,
}) {
  if (attributes == null || attributes.isEmpty) {
    return Ok<TelemetryAttributes>(emptyTelemetryAttributes);
  }
  for (final MapEntry<String, Object?> entry in attributes.entries) {
    if (entry.key.trim().isEmpty || entry.key.codeUnits.contains(0)) {
      return invalidInput<TelemetryAttributes>(
        port: port,
        operation: operation,
        field: 'attributes',
        message: 'Telemetry attribute names must be non-blank and NUL-free',
      );
    }
    final Object? value = entry.value;
    final bool finitePrimitive =
        value is bool || value is String || (value is num && value.isFinite);
    if (!finitePrimitive) {
      return invalidInput<TelemetryAttributes>(
        port: port,
        operation: operation,
        field: 'attributes.${entry.key}',
        message: 'Telemetry attribute values must be finite primitives',
      );
    }
  }
  final List<String> keys = attributes.keys.toList()..sort();
  return Ok<TelemetryAttributes>(
    UnmodifiableMapView<String, Object?>(<String, Object?>{
      for (final String key in keys) key: attributes[key],
    }),
  );
}
